File: src/main/java/dev/isxander/lunarspoof/websocket/asset/LunarAssetWebSocket.java
package dev.isxander.lunarspoof.websocket.asset;

import dev.isxander.lunarspoof.utils.LunarTimer;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft_6455;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.net.URISyntaxException;
import java.nio.ByteBuffer;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

@SuppressWarnings("all")
public class LunarAssetWebSocket extends WebSocketClient {

    public static final Logger LOGGER = LogManager.getLogger("LS Assets");

    public Cache<UUID, Boolean> someCache;
    public LunarTimer someTimer;
    public AssetState currentState;
    public long time;

    public LunarAssetWebSocket(Map<String, String> httpHeaders) throws URISyntaxException {
        super(new URI("wss://assetserver.lunarclientprod.com/connect"), new Draft_6455(), httpHeaders, 30000);
        this.someCache = CacheBuilder.newBuilder().expireAfterWrite(3L, TimeUnit.MINUTES).build();
        this.currentState = AssetState.DISCONNECTED;
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        LOGGER.info("Connection Opened.");

        this.someTimer = null;
        this.currentState = AssetState.AWAITING_ENCRYPTION_REQUEST;
        this.time = System.currentTimeMillis();
    }

    @Override
    public void onMessage(String message) {

    }

    @Override
    public void onMessage(ByteBuffer bytes) {

    }

    private void processMessage() {
        
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        LOGGER.info(String.format("Connection Closed (%d, \"%s\")", code, reason));
    }

    @Override
    public void onError(Exception e) {
        e.printStackTrace();
    }

}

File: src/main/kotlin/dev/isxander/lunarspoof/websocket/asset/LunarAssetWebSocket.kt
package dev.isxander.lunarspoof.websocket.asset

import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import dev.isxander.lunarspoof.utils.LunarTimer
import dev.isxander.lunarspoof.websocket.asset.packet.AbstractWebSocketPacket
import dev.isxander.lunarspoof.websocket.asset.packet.impl.WSPacketClientUnknownCosmetic
import io.netty.buffer.Unpooled
import net.minecraft.network.PacketBuffer
import org.apache.logging.log4j.LogManager
import org.java_websocket.client.WebSocketClient
import org.java_websocket.drafts.Draft_6455
import org.java_websocket.handshake.ServerHandshake
import org.lwjgl.Sys
import java.net.URI
import java.nio.ByteBuffer
import java.util.*
import java.util.concurrent.TimeUnit

@Suppress("unstable")
class LunarAssetWebSocket(
    httpHeaders: Map<String, String>
) : WebSocketClient(
    URI("wss://assetserver.lunarclientprod.com/connect"),
    Draft_6455(),
    httpHeaders,
    30000
) {
    val someCache: Cache<UUID, Boolean> = CacheBuilder.newBuilder().expireAfterWrite(3L, TimeUnit.MINUTES).build()
    var someTimer: LunarTimer? = null
    var currentState: AssetState = AssetState.DISCONNECTED
    var time: Long = 0
    override fun onOpen(handshakedata: ServerHandshake) {
        LOGGER.info("Connection Opened.")
        someTimer = null
        currentState = AssetState.AWAITING_ENCRYPTION_REQUEST
        time = System.currentTimeMillis()
    }

    override fun onMessage(message: String) {}
    override fun onMessage(bytes: ByteBuffer) {
        processPacketBuffer(PacketBuffer(Unpooled.wrappedBuffer(bytes.array())))
    }

    private fun processPacketBuffer(buf: PacketBuffer) {
        val clazz = AbstractWebSocketPacket.REGISTRY.inverse()[buf.readVarIntFromBuffer()]
        try {
            val packet = clazz?.newInstance() ?: return
            LOGGER.debug("Recieved: ${clazz.simpleName}")
            packet.processBuf(buf)
            packet.processSocket(this)
        } catch (e: Exception) {
            LOGGER.error("Error from: $clazz")
            e.printStackTrace()
        }
    }

    fun processUnknownCosmetic(packet: WSPacketClientUnknownCosmetic) {

    }

    override fun onClose(code: Int, reason: String, remote: Boolean) {
        LOGGER.info(String.format("Connection Closed (%d, \"%s\")", code, reason))
    }

    override fun onError(e: Exception) {
        e.printStackTrace()
    }

    companion object {
        val LOGGER = LogManager.getLogger("LS Assets")
    }
}
